gpflow/posteriors.py

```diff
@@ -180,6 +181,121 @@ def _conditional_with_precompute(
+                )
+            except AttributeError:
+                raise ValueError(
+                    "You must pass precompute_cache explicitly (the cache had not been updated before)."
+                )
+        else:
+            self._precompute_cache = precompute_cache
+
+        if precompute_cache is PrecomputeCacheType.NOCACHE:
+            self.alpha = self.Qinv = None
+
+        elif precompute_cache is PrecomputeCacheType.TENSOR:
+            self.alpha, self.Qinv = self._precompute()
+
+        elif precompute_cache is PrecomputeCacheType.VARIABLE:
+            alpha, Qinv = self._precompute()
+            if isinstance(self.alpha, tf.Variable) and isinstance(self.Qinv, tf.Variable):
+                # re-use existing variables
+                self.alpha.assign(alpha)
+                self.Qinv.assign(Qinv)
+            else:  # create variables
+                self.alpha = tf.Variable(alpha, trainable=False)
+                self.Qinv = tf.Variable(Qinv, trainable=False)
+
+
+class GPRPosterior(AbstractPosterior):
+    def __init__(
+        self,
+        kernel,
+        X_data: tf.Tensor,
+        Y_data: tf.Tensor,
+        likelihood_variance: Parameter,
+        mean_function: Optional[mean_functions.MeanFunction] = None,
+        *,
+        precompute_cache: Optional[PrecomputeCacheType],
+    ):
+
+        super().__init__(kernel, X_data, mean_function=mean_function)
+        self.mean_function = mean_function
+        self.Y_data = Y_data
+        self.likelihood_variance = likelihood_variance
+
+        if precompute_cache is not None:
+            self.update_cache(precompute_cache)
+
+    def _add_noise_cov(self, K: tf.Tensor) -> tf.Tensor:
+        """
+        Returns K + σ² I, where σ² is the likelihood noise variance (scalar),
+        and I is the corresponding identity matrix.
+        """
+        # taken straight from deprecated version
+        k_diag = tf.linalg.diag_part(K)
+        s_diag = tf.fill(tf.shape(k_diag), self.likelihood_variance)
+        return tf.linalg.set_diag(K, k_diag + s_diag)
+
+    def _conditional_with_precompute(
+        self, Xnew, full_cov: bool = False, full_output_cov: bool = False
+    ) -> MeanAndVariance:
+        """
+        Computes predictive mean and (co)variance at Xnew, *excluding* mean_function.
+        Relies on cached alpha and Qinv.
+        """
+        Kmn = self.kernel(self.X_data, Xnew)
+        Knn = self.kernel(Xnew, full_cov=full_cov)
+
+        return base_conditional_with_lm(Kmn, self.Qinv, Knn, self.alpha, full_cov=full_cov)
+
+    def _precompute(self) -> Tuple[tf.Tensor, tf.Tensor]:
+
+        Kmm = self.kernel(self.X_data)
+        Kmm_plus_s = self._add_noise_cov(Kmm)
+
+        # we cache the cholesky decomposition of Kmm_plus_s because we will call
+        # base_conditional_with_lm implementation
+        Lm = tf.linalg.cholesky(Kmm_plus_s)
+
+        alpha = self.Y_data - self.mean_function(self.X_data)  # type: ignore
```

Comment:
What is the problem with the types here?

Reply:
It is because the mean_function default value is None (consistent with other posterior implementations), and so it throws up a 'None' not callable here.  

But we never have to worry about a None mean_function [because of this] (https://github.com/Andrew878/GPflow/blob/b4e6298c295c6ca90f04ecd69cceeccf450b59ea/gpflow/models/model.py#L108) in the GPModel parent class a few levels up

I don't normally like adding 'ignores' so if you know a better way, I am happy to learn and fix it! (one idea I had is having a `Zero()` mean function as the default value)